api/src/main/java/org/pmiops/workbench/db/dao/UserService.java

```diff
@@ -73,6 +71,30 @@ private void updateDataAccessLevel(User user) {
     }
   }
 
+  public User createUser(String givenName, String familyName, String email, String contactEmail) {
+    User user = new User();
+    user.setDataAccessLevel(DataAccessLevel.UNREGISTERED);
+    user.setEmail(email);
+    user.setContactEmail(contactEmail);
+    user.setFamilyName(familyName);
+    user.setGivenName(givenName);
+    try {
+      userDao.save(user);
+    } catch (DataIntegrityViolationException e) {
+      user = userDao.findUserByEmail(email);
+      if (user == null) {
+        throw e;
+      }
+      // If a user already existed, update it in place.
```

Comment:
What case is this meant to cover? In what valid scenario would a user already exist (can you add that to the comment)? Would someone be able to use this to go through the signup process again and mutate their user details?

Reply:
It was specifically for the case where the user signs in without a user record, and multiple requests simultaneously try to create one.

I realized that in this case it shouldn't be necessary to update the existing user... and it seems safer not to, so I removed that logic.

Added a comment.